### PR TITLE
feature/mip_test

### PR DESCRIPTION
**Pull request by xross**
Add functionally to run tests with other cores running (dummy threads)

Also improved test code re-use

### CODE REVIEW

**Comment by xross** [comment]
All checks now passing :)

File: tests/shared_src/shared.h [hunk 1]
```diff
@@ -200,22 +205,72 @@ int TestEp_Rx(chanend c_out, int epNum, int start, int end)
...
 {
+#if (DUMMY_THREAD_COUNT > 0)
     par(size_t i = 0; i < DUMMY_THREAD_COUNT; i++)
     {
-        dummyThread(c[i]);
+        dummyThread();
     }
+#endif
 }
+
```
Comment (mbanth) [inline tests/shared_src/shared.h:270]: The number of blank lines seems a bit excessive.

File: tests/shared_src/test_main.xc [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+int main()
+#endif
+{
+    chan c_ep_out[EP_COUNT_OUT], c_ep_in[EP_COUNT_IN];
+
+    par
+    {
+        {
+#if defined(XUD_TEST_SPEED_FS)
+            unsigned speed = XUD_SPEED_FS;
```
Comment (mbanth) [inline tests/shared_src/test_main.xc:16]: If speed is constant, define as `const`.

File: tests/test_bulk_rx_multiep/src/main.xc [hunk 3]
```diff
@@ -25,28 +25,22 @@
...
-        }
+        fail[0] = TestEp_Rx(c_ep_out[TEST_EP_NUM], TEST_EP_NUM, PACKET_LEN_START, PACKET_LEN_END);
+        fail[1] = TestEp_Rx(c_ep_out[4], 4, PACKET_LEN_START, PACKET_LEN_END);
+        fail[2] = TestEp_Rx(c_ep_out[5], 5, PACKET_LEN_START, PACKET_LEN_END);
+        fail[3] = TestEp_Rx(c_ep_out[6], 6, PACKET_LEN_START, PACKET_LEN_END);
     }
 
-    return 0;
+    for(size_t i = 0; i< 4; i++)
+        fail[0] |= fail[i];
```
Comment (mbanth) [inline tests/test_bulk_rx_multiep/src/main.xc:41]: How does the engineer investigating a failure tell which run of `TestEp_Rx()` failed?

File: tests/test_bulk_tx_badack/src/main.xc [hunk 4]
```diff
@@ -13,58 +12,27 @@
 #define EP_COUNT_OUT          (6)
 #define EP_COUNT_IN           (6)
 
-/* Endpoint type tables */
+#include "shared.h"
+
 XUD_EpType epTypeTableOut[EP_COUNT_OUT] = {XUD_EPTYPE_CTL,
                                                 XUD_EPTYPE_BUL,
```
Comment (mbanth) [inline tests/test_bulk_tx_badack/src/main.xc:18]: It appears that this line has a different indentation to the ones which follow.  Is that intentional?

File: tests/test_bulk_tx_basic/src/main.xc [hunk 5]
```diff
@@ -13,7 +12,8 @@
 #define EP_COUNT_OUT   		(6)
 #define EP_COUNT_IN    		(6)
 
-/* Endpoint type tables */
+#include "shared.h"
+
 XUD_EpType epTypeTableOut[EP_COUNT_OUT] = {XUD_EPTYPE_CTL,
                                                 XUD_EPTYPE_BUL,
                                                  XUD_EPTYPE_BUL,
```
Comment (mbanth) [inline tests/test_bulk_tx_basic/src/main.xc:19]: It appears that this line has a different indentation to the ones surrounding it.  Is that intentional?

File: tests/test_bulk_tx_multiep/src/main.xc [hunk 6]
```diff
@@ -1,47 +1,46 @@
...
-#define PKT_LENGTH_START 10
-#define PKT_LENGTH_END 19
+#define PKT_LENGTH_START   (10)
+#define PKT_LENGTH_END     (19)
 
+#include "shared.h"
 
-int main()
+XUD_EpType epTypeTableOut[EP_COUNT_OUT] = {XUD_EPTYPE_CTL,
+                                                XUD_EPTYPE_BUL,
```
Comment (mbanth) [inline tests/test_bulk_tx_multiep/src/main.xc:13]: It appears that this line has a different level of indentation that the ones which follow it.  Is that intentional?

File: tests/test_bulk_tx_multiep/src/main.xc [hunk 7]
```diff
@@ -1,47 +1,46 @@
...
-
+        fail[0] = TestEp_Tx(c_ep_in[3], 3, PKT_LENGTH_START, PKT_LENGTH_END, RUNMODE_DIE);
+        fail[1] = TestEp_Tx(c_ep_in[4], 4, PKT_LENGTH_START, PKT_LENGTH_END, RUNMODE_DIE);
+        fail[2] = TestEp_Tx(c_ep_in[5], 5, PKT_LENGTH_START, PKT_LENGTH_END, RUNMODE_DIE);
+        fail[3] = TestEp_Tx(c_ep_in[6], 6, PKT_LENGTH_START, PKT_LENGTH_END, RUNMODE_DIE);
     }
 
-    return 0;
+    for(size_t i = 0; i < 4; i++)
+        fail[0] |= fail[i];
```
Comment (mbanth) [inline tests/test_bulk_tx_multiep/src/main.xc:39]: How does the engineer investigating a failure of this test determine which run of `TestEp_Tx()` failed?

File: tests/test_bulk_tx_multiep/src/main.xc [hunk 8]
```diff
@@ -1,47 +1,46 @@
...
-#include "shared.h"
-
-#define XUD_EP_COUNT_OUT   5
-#define XUD_EP_COUNT_IN    8
 
-/* Endpoint type tables */
-XUD_EpType epTypeTableOut[XUD_EP_COUNT_OUT] = {XUD_EPTYPE_CTL, XUD_EPTYPE_BUL, XUD_EPTYPE_BUL, XUD_EPTYPE_BUL, XUD_EPTYPE_BUL};
-XUD_EpType epTypeTableIn[XUD_EP_COUNT_IN] =   {XUD_EPTYPE_CTL, XUD_EPTYPE_BUL, XUD_EPTYPE_BUL, XUD_EPTYPE_BUL, XUD_EPTYPE_BUL, 
-                                               XUD_EPTYPE_BUL, XUD_EPTYPE_BUL, XUD_EPTYPE_BUL};
+#define EP_COUNT_OUT       (4)
```
Comment (mbanth) [inline tests/test_bulk_tx_multiep/src/main.xc:4]: Why the reduction of End Points from 5 to 4?

File: tests/test_iso_tx_basic/src/main.xc [hunk 9]
```diff
@@ -11,45 +9,37 @@
...
+
+#include "shared.h"
+
 XUD_EpType epTypeTableOut[EP_COUNT_OUT] = {XUD_EPTYPE_CTL,
-                                                XUD_EPTYPE_BUL,
-                                                 XUD_EPTYPE_BUL,
-                                                 XUD_EPTYPE_BUL,
-                                                 XUD_EPTYPE_BUL,
-                                                 XUD_EPTYPE_BUL};
+                                                XUD_EPTYPE_ISO,
```
Comment (mbanth) [inline tests/test_iso_tx_basic/src/main.xc:23]: THis line appears to have a different level of indentation than the ones following.  Is that intentional?

**Review by mbanth — CHANGES_REQUESTED** [review]

File: tests/shared_src/test_main.xc (hunk 2 above)
Comment (xross) [reply tests/shared_src/test_main.xc:16]: Fixed

File: tests/test_bulk_rx_multiep/src/main.xc (hunk 3 above)
Comment (xross) [reply tests/test_bulk_rx_multiep/src/main.xc:41]: They would have to add debug code. Currently all tests are pass/fail only - no other feedback.

File: tests/test_bulk_tx_badack/src/main.xc (hunk 4 above)
Comment (xross) [reply tests/test_bulk_tx_badack/src/main.xc:18]: Fixed.

File: tests/test_bulk_tx_basic/src/main.xc (hunk 5 above)
Comment (xross) [reply tests/test_bulk_tx_basic/src/main.xc:19]: Fixed

File: tests/test_bulk_tx_multiep/src/main.xc (hunk 7 above)
Comment (xross) [reply tests/test_bulk_tx_multiep/src/main.xc:39]: Fixed

File: tests/test_bulk_tx_multiep/src/main.xc (hunk 8 above)
Comment (xross) [reply tests/test_bulk_tx_multiep/src/main.xc:4]: Changed back to 5, doesn't really matter as they are not used in this test.

File: tests/test_iso_tx_basic/src/main.xc (hunk 9 above)
Comment (xross) [reply tests/test_iso_tx_basic/src/main.xc:23]: Fixed

File: tests/test_bulk_tx_multiep/src/main.xc (hunk 6 above)
Comment (xross) [reply tests/test_bulk_tx_multiep/src/main.xc:13]: Fixed

File: tests/test_bulk_tx_multiep/src/main.xc (hunk 7 above)
Comment (xross) [reply tests/test_bulk_tx_multiep/src/main.xc:39]: See previous comment

File: tests/test_bulk_rx_multiep/src/main.xc (hunk 3 above)
Comment (mbanth) [reply tests/test_bulk_rx_multiep/src/main.xc:41]: I understand, and that's okay with me.

Comment (xross) [reply tests/test_bulk_rx_multiep/src/main.xc:41]: I agree that it it would be nice if the test framework did output better failure reasoning.

**Comment by xross** [comment]
@mbanth is this okay now please?

File: tests/shared_src/test_main.xc [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+                unsigned * unsafe p = &g_dummyThreadCount;
+                *p = 0;
+            }
+            if(TEST_EP_NUM != 0)
+            {
+                XUD_ep ep_out_0 = XUD_InitEp(c_ep_out[0]);
+			    XUD_Kill(ep_out_0);
+			}
+            exit(0);
+		}
```
Comment (DaveAtkinsonXMOS) [inline tests/shared_src/test_main.xc:47]: Something gone a bit awry with the indentation here?

**Review by mbanth — APPROVED** [review]
All looks good.

File: tests/test_bulk_tx_basic/src/main.xc [hunk 11]
```diff
@@ -13,59 +12,28 @@
...
+                                                XUD_EPTYPE_BUL,
+                                                XUD_EPTYPE_BUL,
+                                                XUD_EPTYPE_BUL,
+                                                XUD_EPTYPE_BUL};
 XUD_EpType epTypeTableIn[EP_COUNT_IN] =   {XUD_EPTYPE_CTL, 
                                                 XUD_EPTYPE_BUL,
                                                 XUD_EPTYPE_BUL,
                                                 XUD_EPTYPE_BUL,
                                                 XUD_EPTYPE_BUL,
                                                 XUD_EPTYPE_BUL};
```
Comment (DaveAtkinsonXMOS) [inline tests/test_bulk_tx_basic/src/main.xc:29]: Following on from Michael's observation about indentation, the above is similare...

**Review by DaveAtkinsonXMOS — APPROVED** [review]
LGTM

File: tests/shared_src/test_main.xc (hunk 10 above)
Comment (xross) [reply tests/shared_src/test_main.xc:47]: Will fix

**Pull request merged**